text/0567-use-and-resources.md

```diff
@@ -0,0 +1,1014 @@
+  });
+}
+```
+
+Generators are a powerful syntax, but currently they do not have lexical binding
+built in (a la arrow functions). Passing the parent to the manager prevents
+having to add additional boilerplate for their usage. The default resource
+implementation will not directly expose the parent, however.
+
+Both `setupResource` and `updateResource` should return an object that has a
+`state` getter on it. This allows `@use` and `{{use}}` to maintain a reference
+to the internal state of the resource, and get it again whenever it updates. The
+timing semantics of the hooks are as follows:
+
+#### `createResource`
+
+- Called to create the resource.
+
+#### `getState`
+
+- Called whenever the state of the resource is accessed.
+
+#### `setupResource`
+
+- Called immediately after the resource is created (using `useResource`).
+
+#### `updateResource`
+
+- Called after the arguments to the resource have changed, or after any
+  values tracked during the previous `setupResource` or `updateResource`
+  have changed (unless the `disableAutotracking` capability is set to true).
+- Called asynchronously by default, unless the resource is accessed, in
+  which case it is called synchronously before returning its internal state.
+- Called before the equivalent asynchronous update resources of any parent:
+  - Components
+  - Modifiers
+  - Resources
+
+#### `destroyResource`
+
+- Called asynchronously after the instance of the parent class that the resource
+  is defined on is scheduled for destruction.
+- Called _before_ the equivalent asynchronous destruction resources of any
+  parent:
+  - Components
+  - Modifiers
+  - Resources
+
+Resource managers have the following capabilities and defaults:
+
+```ts
+interface ResourceManagerCapabilities {
+  disableAutotracking: boolean = false;
+}
+```
+
+The effects of these capabilities are as follows:
+
+- `disableAutotracking`: If set to false, then the `setupResource` and
+  `updateResource` hooks will be autotracked, and changes to the state consumed
+  while running them will trigger the `updateResource` hook again. If set to
+  true, only changes to the incoming arguments themselves will cause updates.
+
+Resource managers can be set on classes or objects using an API similar to the
+component and modifier manager APIs:
+
+```js
+setResourceManager(
+  owner => new CustomResourceManager(owner),
+  CustomResource
+);
+```
+
+Like other managers, this manager will inherit down the prototype hierarchy, so
+it only needs to be set on the base class.
+
+### `useResource`
+
+`useResource` is responsible for creating a resource and associating it with a
+parent, such as a component, controller, or another resource. It receives two
+arguments, and returns a resource instance.
+
+```ts
+type ManagedResource = object;
+type ResourceDefinition = { resource: ManagedResource };
+type ResourceDefinitionThunk = () => ResourceDefinition;
+
+interface ResourceInstance {
```

Comment:
ZOMG LETS NOT INCLUDE INSTANCE IN AN INTERFACE / CLASS NAME AGAIN PLZ?!?!?!!?!

😸 ❤️

Reply:
We should come up with a name not as apt to troll you for sure. 😆 

I’d suggest we invert this: it should be `interface Resource` and if we need to name the type of the resource constructor, differentiate *that* as `interface ResourceConstructor`. We *may* need to make that distinction here because of how TS conceptualizes its types, i.e. that you *do* have to distinguish between the type and it’s constructor. However, as this is designed currently, I don’t *think* we need that: we can just specify that they type for the argument to `resource` is a `Resource`, full stop?

(I’ll have more comments on the types proposed here later this week. @pzuraq and @dfreeman and I have played with a bunch of variants in this space and the types as they are in the proposal here represent a fairly rough draft which I’d like to tighten down and polish up so that what the Typed Ember crowd ends up doing on DefinitelyTyped can actually match the design here. Otherwise we’re asking for confusion in the community, and nobody wants that! 😂)